Функциональность Поиск.

AleksandrTimofeev1987 opened the pull request:
feature: добавить end-point GET /fimls/search
feature: добавить метод FilmService.searchFilmsByTitleOrDirector
feature: добавить метод DbFilmStorage.searchFilmsByTitle
feature: добавить метод DbFilmStorage.searchFilmsByDirector
feature: добавить метод DbFilmStorage.searchFilmsByTitleAndDirector

AleksandrTimofeev1987 committed:
Функциональность Поиск.
feature: добавить end-point GET /fimls/search
feature: добавить метод FilmService.searchFilmsByTitleOrDirector
feature: добавить метод DbFilmStorage.searchFilmsByTitle
feature: добавить метод DbFilmStorage.searchFilmsByDirector
feature: добавить метод DbFilmStorage.searchFilmsByTitleAndDirector

BucketOnHead commented on src/main/java/ru/yandex/practicum/filmorate/exception/IncorrectSearchParameterException.java:
```diff
@@ -0,0 +1,8 @@
+package ru.yandex.practicum.filmorate.exception;
+
+public class IncorrectSearchParameterException extends RuntimeException {
+
```
Пустая строка)

BucketOnHead reviewed (approved):
Я не увидел злостных нарушений, смог прицепиться только к пустой строке) Передаю Approve

Vintros commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -64,5 +65,18 @@ public void removeFilmById(@PathVariable Long filmId) {
+        return filmService.searchFilmsByTitleOrDirector(query, by);
+    }
+
+    private static void validateSearchParameter(String by) {
```
Валидацию лучше перенести в класс валидатор, а саму валидность проверять в сервисе(если это не проверка аннотациями ломбока).

AleksandrTimofeev1987 committed:
Merge branch 'develop' into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java

AleksandrTimofeev1987 committed:
Функциональность Поиск.
refactor: перенести валидацию поискового запроса в класс Validator
refactor: перенести инициализацию валидации поискового запроса в FilmService.searchFilmsByTitleOrDirector

Vintros commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java:
```diff
@@ -306,5 +309,85 @@ public List<Film> findPopularFilmSortedByGenreAndYear(int count, long genreId, i
+        // Получаем список данных Имя директора + ID фильма
+        final String sqlQuery = "SELECT DIR.NAME AS DIRECTOR_NAME, DIRS.FILM_ID FROM DIRECTORS AS DIRS " +
+                "LEFT JOIN DIRECTOR AS DIR ON DIRS.DIRECTOR_ID = DIR.DIRECTOR_ID";
+        List<Map.Entry<Long, String>> dataList = jdbcTemplate.query(sqlQuery, this::mapRowToMapEntryFilmIdDirectorName);
```
Мапу в лист не обязательно оборачивать, она же потом все равно назад извлекается из листа. Достаточно в методе mapRowToMapEntryFilmIdDirectorName из параметров убрать (int i) и тогда сформируется сразу мапа

AleksandrTimofeev1987 replied:
Так я не мапу извлекаю, а entry. Если убрать параметр (int i) то на выходе я получу один экземпляр entry, хотя записей у меня много. А оборачивать entry в лист или сразу обрабатывать особой разницы нет, все равно каждый entry проверять придется. Или я тебя не так понял?

Vintros replied:
Да, в коде все правильно, это я немного затупил ). Замечание снимается

Vintros commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java:
```diff
@@ -306,5 +309,85 @@ public List<Film> findPopularFilmSortedByGenreAndYear(int count, long genreId, i
+
+    private List<Film> getFilmsSortedByPopularity(List<Long> matchingIds) {
+        return matchingIds.stream()
+                .map(this::getFilmById)
```
Запрос каждого фильма из БД отдельным методом это так называемая проблема n+1, когда увеличение количества объектов пропорционально увеличивает количество запросов к БД, этого нужно максимально избегать, количество запросов к БД не должно зависеть от количества запрашиваемых объектов.
Как вариант: заново передать список айдишников в sql запрос фильмов, в параметр WHERE и получить готовый список фильмов одним запросом, сразу отсортированный по популярности.
Второй вариант: сначала запросить список всех фильмов уже имеющимся методом getFilms() и из полученного списка отфильтровать нужные фильмы.

AleksandrTimofeev1987 replied:
Поправлю

Vintros commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -64,5 +65,18 @@ public void removeFilmById(@PathVariable Long filmId) {
     }
+
+    @GetMapping("/search")
+    public List<Film> searchFilmsByTitleOrDirector(@RequestParam String query,
```
Не рассмотрен вариант из ТЗ, где отсутствуют параметры запроса и нужно вернуть все фильмы.

Vintros commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -64,5 +65,18 @@ public void removeFilmById(@PathVariable Long filmId) {
+
+    @GetMapping("/search")
+    public List<Film> searchFilmsByTitleOrDirector(@RequestParam String query,
+                                                @RequestParam(defaultValue = "title") String by) {
```
А зачем здесь defaultValue? Мне кажется если параметр by не передан вообще - нужно кидать эксепшен, а не додумывать за пользователя, что он хочет искать именно по названию, а например не по режиссеру.

AleksandrTimofeev1987 replied:
Поставил в самом начале. Забыл убрать, спасибо, что заметил.)

Vintros reviewed (changes requested):
Несколько замечаний написал )

AleksandrTimofeev1987 committed:
Функциональность Поиск. Корректировка после рувью.
refactor: переписать методы поиска в DbFilmStorage, чтобы избежать множественных обращений к БД

BucketOnHead commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -125,14 +125,22 @@ public List<Film> getFilmsByDirector(Long directorId, String sortBy) {
     public List<Film> searchFilmsByTitleOrDirector(String query, String by) {
```
Предлагаю переименовать by на searchBy для большей ясности)

AleksandrTimofeev1987 committed:
Функциональность Поиск. Корректировка после ревью.
style: переименовать параметр by на searchBy у метода FilmService.searchFilmsByTitleOrDirector
fix: обратить сортировку в методе DbFilmStorage.populateFilmsWithLikes

Vintros commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -117,7 +119,28 @@ public void removeFilmById(Long id) {
+    }
+
+    public List<Film> searchFilmsByTitleOrDirector(String query, String searchBy) {
+        log.info("Получен поисковый запрос: {}. Парамент поиска: {}", query, searchBy);
```
Очепятка: Парамент поиска

Vintros commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java:
```diff
@@ -318,5 +319,110 @@ public List<Film> getCommonFilms(Long userId, Long friendId) {
+    }
+
+    private List<Film> getFilmsSortedByPopularity(List<Long> matchingIds) {
+        String inSql = String.join(",", Collections.nCopies(matchingIds.size(), "?"));
```
+++

AleksandrTimofeev1987 committed:
Функциональность Поиск. Корректировка после ревью.
style: исправить опечатку в логировании метода FilmService.searchFilmsByTitleOrDirector

Vintros reviewed (approved):
Апрув!
